hamt/hamt_test.go

```diff
@@ -288,6 +291,9 @@ func TestSetAfterMarshal(t *testing.T) {
 		}
 	}
 
+	// fails
+	nds.nd.Cid()
```

Comment:
@schomatis

We noticed something and want to get some feedback. This line `nds.nd.Cid()` results in a `nil exception` because while trying to calculate the `cid` some of the links are `nil`. We ran into this, because we need to pass the `cid` to @whyrusleeping's `EnumerateChildrenAsync` method. We have a question:

This behavior is not new. If you go to master and put the same line in the same place it will also result in the same error. Which we've determined is because calling  `nds.Set` does NOT populate the `nds.nd.links`, but it does add `nds.children`. This was not an issue before because `EnumLinks` walked the children and not the links, but this new solution deals only with `ProtoNode` and not `Shard`. Is this ok? Does this test reflect something that another developer in the future might do: i.e. create a shard, add a bunch of nodes, and then run `EnumLinks` on it?

Reply:
Honestly, I have no idea. The only people that know what's going on here are @whyrusleeping and @Stebalien.

Let me give you my take from an outside developer who's trying to figure out the logic here with the few comments I see on this package.

Yes, inserting a new link in the directory only saves it in the `children` slice of shards but not in the node links where it's just placing a `nil` placeholder,

https://github.com/ipfs/go-unixfs/blob/06009a2f87105928776bc9994cd712b034df01bd/hamt/hamt.go#L341-L342

From the `getChild` comments and logic I'm inferring that the `children` slice is just a cache of the links of the node in `nd` which have the "actual truth" of the state of the HAMT,

https://github.com/ipfs/go-unixfs/blob/06009a2f87105928776bc9994cd712b034df01bd/hamt/hamt.go#L256-L259

since when the links of the internal node are being accessed they are *never* checked for `nil` values (it's always assumed that you can read the link `Name`).

https://github.com/ipfs/go-unixfs/blob/06009a2f87105928776bc9994cd712b034df01bd/hamt/hamt.go#L276-L280

The `Node` method does seem to guarantee a node that will have link integrity since it contemplates accessing both slices depending on the case (something about a "child unloaded" case I'm not really sure what it's referring to),

https://github.com/ipfs/go-unixfs/blob/06009a2f87105928776bc9994cd712b034df01bd/hamt/hamt.go#L152-L173

that doesn't panic in my tests but I guess it's very expensive to be computing it in every `EnumLinks` call.

Also, I'm not sure if the internal `ProtoNode` of the shard (`nd`) should be trusted as a faithful representations of the shard at every moment of its operation. From my (very limited) understanding `ProtoNode`s are not mutable objects, or at least they should be handled with care if they want to be used to that end, I would normally advise to make a copy of the node and work on that. The `Cid` method itself is very dangerous since it relies on [cached versions](https://github.com/ipfs/go-merkledag/blob/master/node.go#L309) of the computed hash, a cache that is not always invalidated when the node is modified, e.g., the [`SetLinks`](https://github.com/ipfs/go-merkledag/blob/master/node.go#L347) method used here doesn't do it (the same with the weird link copy operations in `rmChild`) compared with methods like [`AddRawLink`](https://github.com/ipfs/go-merkledag/blob/master/node.go#L116) which does invalidate it. As a simple example, if you add a `Cid` call to the beginning of any of these tests (just after the root shard creation), following calls won't panic (because they actually won't access the node links, just its cached hash).

Anyway, my answer is still "I'm not really sure", just wanted to give you some background here.

(Btw, we really need to shape up the code in this package, big time. Any external audit of this package -and by external I just mean anyone besides the three developers that have done any commits to the `hamt.go` file- is really a very expensive operation. My first impression when I landed in the `TestSetAfterMarshal` function and saw the three variables named `nd`, `ds`, and `nds` was of anxiety, and it goes downhill from there.. This shouldn't be so hard as anyone can grasp the general idea of redistributing link entries in many nodes instead of overloading just one, but I feel that the implementation details are somewhat obscure, `maxpadlen`? `makeDirWidth` talks about size *and* width.. I'll open a separate issue about this, sorry for the rant :)

Reply:
@schomatis thank you for a thorough and thoughtful response, I'm processing it all now

Reply:
Unfortunately, I'm also not sure what's going on here. I'm guessing this has something to do with lazily calculating everything when we call `Shard.Node()` but I don't see us actually *doing* that.

Reply:
> I'm guessing this has something to do with lazily calculating everything when we call `Shard.Node()`

Yes, I think the safest bet at this point would be to iterate the links in the way `Node()` does,

https://github.com/ipfs/go-unixfs/blob/3738990784c63b9bfa2fc104c06c50ece6729aef/hamt/hamt.go#L149-L178

(without the expensive `DAGService.Add` call or marshaling the data) which could be encapsulated in a function to avoid repeating code.